### ISSUE TITLE
`Unrecognized brand : different at top`, when there is really no difference

### ISSUE THREAD

**Issue by vporton** [Agoric/agoric-sdk#986]
https://github.com/vporton/time-release/tree/823aa8258a6ed41cfc974b061e508da3248e1cf8

`yarn test` produces this error. Apparently at the line https://github.com/vporton/time-release/blob/823aa8258a6ed41cfc974b061e508da3248e1cf8/contract/src/contracts/proxy.js#L86

`amount` and `payment` have the same brand, but apparently are detected as different brands.

**Comment by DavidBruant** [comment]
I did:
```sh
git clone git@github.com:vporton/time-release.git
cd time-release/
git checkout 823aa82
agoric install
agoric start
yarn test
```

and I can reproduce the issue

I'm investigating

**Comment by DavidBruant** [comment]
Full error stack trace:
```
LOGGED ERROR: Unrecognized brand : different at top : ( {
  isMyIssuer: [Function: isMyIssuer],
  getAllegedName: [Function: getAllegedName]
} ) vs ( {
  isMyIssuer: [Function: isMyIssuer],
  getAllegedName: [Function: getAllegedName]
} ) Error: Unrecognized brand: different at top: ((an object)) vs ((an object))
See console for error data.
    at Object.complain (/home/david/projects/agoric-sdk/packages/assert/src/assert.js:139:19)
    at Function.fail (/home/david/projects/agoric-sdk/packages/assert/src/assert.js:163:20)
    at complain (/home/david/projects/agoric-sdk/packages/same-structure/src/sameStructure.js:168:12)
    at mustBeSameStructureInternal (/home/david/projects/agoric-sdk/packages/same-structure/src/sameStructure.js:197:9)
    at mustBeSameStructure (/home/david/projects/agoric-sdk/packages/same-structure/src/sameStructure.js:239:3)
    at Object.coerce (/home/david/projects/agoric-sdk/packages/ERTP/src/amountMath.js:189:7)
    at /home/david/projects/agoric-sdk/packages/zoe/src/cleanProposal.js:46:38
    at Array.map (<anonymous>)
    at coerceAmountKeywordRecord (/home/david/projects/agoric-sdk/packages/zoe/src/cleanProposal.js:45:41)
    at cleanProposal (/home/david/projects/agoric-sdk/packages/zoe/src/cleanProposal.js:98:10)
```
(repeated twice)

---

If i change
`console.log(amount.brand.getAllegedName(), payment.getAllegedBrand().getAllegedName())`
to
`console.log('brands', amount.brand.getAllegedName(), payment.getAllegedBrand().getAllegedName(), amount.brand === payment.getAllegedBrand());`, i get:

> brands BaytownBucks BaytownBucks true

So not only the alleged names, but the brand object themselves are identical (it's a more trustworthy test for brands), so the problem may come from somewhere else

I keep investigating

**Comment by DavidBruant** [comment]
Ok, i think i've found the mismatch
The payment and amount have the same brand. However, they are not of the expected brand. 

https://github.com/vporton/time-release/blob/823aa8258a6ed41cfc974b061e508da3248e1cf8/contract/src/contracts/proxy.js#L88

```js
{ give: { Wrapper: amount } }
```

The `amount` should be of the `Wrapper` brand, but isn't.

If i trace back to where this amount comes from, it's from the `purse` which comes from `const purse = await timeLock.getPayment();`

https://github.com/vporton/time-release/blob/823aa8258a6ed41cfc974b061e508da3248e1cf8/contract/src/contracts/proxy.js#L71

The `payment` in `timeLock.getPayment()` comes from the `payments` Map

https://github.com/vporton/time-release/blob/823aa8258a6ed41cfc974b061e508da3248e1cf8/contract/src/contracts/proxy.js#L67

...which is filled with payments coming from the `paymentIssuer` passed as argument to the `sendHook`

```js
    const sendHook = (paymentIssuer, lockedPayment, handle, date) => userOfferHandle => {
      const myPurse = paymentIssuer.makeEmptyPurse();
      const lockedAmount = myPurse.deposit(lockedPayment);

      const lock = makeTimeRelease(zcf, timerService, myPurse, date);

      payments.set(handle, lock);
```

https://github.com/vporton/time-release/blob/823aa8258a6ed41cfc974b061e508da3248e1cf8/contract/src/contracts/proxy.js#L31-L37

and i believe, this is the `BaytownBucks` issuer

```js
const currencyIssuer = produceIssuer('BaytownBucks')
    const { mint: baytownBucksMint, issuer } = currencyIssuer;
    const baytownBucks = issuer.getAmountMath().make;

    const payment = baytownBucksMint.mintPayment(baytownBucks(1000));

    function pushPullMoney(date) {
      const sendInvite = inviteIssuer.claim(
        publicAPI.makeSendInvite(harden(issuer), harden(payment), harden(handle), harden(date))());
```

https://github.com/vporton/time-release/blob/823aa8258a6ed41cfc974b061e508da3248e1cf8/contract/test/test-proxy.js#L59-L67

---

Long story short, in `{ give: { Wrapper: amount } }`, the expected amount is of the `Wrapper` issuer, but it's actually from the `BaytownBucks` issuer, hence the brand mismatch error

I'm not sure how to fix your code, but you can discuss that elsewhere

---

I'm closing this issue since it doesn't seem like there is a bug on the agoric-sdk side

However, i think that the error message is really bad and could be improved mention the 2 alleged brand names being compared. It would have saved you and me hours. So i'm making a followup issue.

Thank you for the bug report :-)
